### PR TITLE
Automated VIC-UI Plugin - Install

### LINKED ISSUE #1702: Automated UI plugin install
**User Statement:**
As a user I want the UI plugin install to be automated.

**Details:**

**Acceptance Criteria:**
- [x] Automated UI plugin install
- [ ] ~Works with `register` API~ - no longer desired.
- [x] Improved logging in `vic-ui` code especially around certs
- [ ] Documentation update
- [ ] Document manual process
- [x] Consider uninstall
- [x] Consider upgrade
- [x] https://github.com/vmware/vic-ui/issues/452

### LINKED ISSUE #1789: Block fileserver responses for favicon.ico
Bug Report Detailed Information
-------------------------------

- What operation was being performed when the failure was noticed?

- Provide information from the `Support Information` section of the appropriate Appliance Lifecycle stage:
  https://github.com/vmware/vic-product/blob/master/installer/docs/SUPPORT.md#appliance-lifecycle

- Provide additional detail as necessary:

The fileserver is currently treating all requests on `/` as an IndexHandler request. 
Consequently, all additional root requests, such as '/favicon.ico', receive the index.html payload.

This situation can be resolved with a NotFound handler on the root http path.

### PR DESCRIPTION
**Pull request by morris-jason**
VIC Appliance Checklist:
- [ ] Up to date with `master` branch
- [ ] Added tests
- [ ] Considered impact to upgrade
- [ ] Tests passing
- [ ] Updated documentation
- [ ] Impact assessment checklist

If this is a feature or change to existing functionality, consider areas of impact with the [Impact
Assessment Checklist](https://github.com/vmware/vic-product/blob/master/installer/docs/CHANGE.md)

Progress to #1702, Fixes #1789

This pr implements plugin lifecycle operations through the index page and through new `/plugin` apis.

Reviewing by commit is probably easiest, as the first few commits are just copying files.

### CODE REVIEW

File: installer/fileserver/tasks/ova/configure.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		ManagedBy: &types.ManagedByInfo{
+			ExtensionKey: ManagedByKey,
+			Type:         ManagedByType,
+		},
+	}
+
+	info, err := v.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
+		return v.Reconfigure(ctx, spec)
+	})
+
```
Comment (jzt) [inline installer/fileserver/tasks/ova/configure.go]: consistency: remove this whitespace

File: installer/fileserver/tasks/ova/configure_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (jzt) [inline installer/fileserver/tasks/ova/configure_test.go]: `Copyright 2016-2018`

File: installer/fileserver/tasks/ova/configure.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (jzt) [inline installer/fileserver/tasks/ova/configure.go]: Copyright `2016-2018`

File: installer/fileserver/routes/plugin.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
```
Comment (jzt) [inline installer/fileserver/routes/plugin.go]: Copyright `2017-2018`

File: installer/fileserver/tasks/plugin.go [hunk 5]
```diff
@@ -44,141 +45,83 @@ type Plugin struct {
...
+
+		if err != nil {
+			log.Error(err.Error())
+			return err
+		}
 	}
 
 	if p.Company == "" {
-		return cli.NewExitError("--company must be specified", 1)
+		return errors.New("company must be specified")
```
Comment (jzt) [inline installer/fileserver/tasks/plugin.go]: Might be cleaner to convert this group of ifs into a switch block for readability/compactness.

File: installer/fileserver/tasks/plugin.go [hunk 6]
```diff
@@ -280,11 +228,13 @@ func (p *Plugin) Install(cli *cli.Context) error {
 	log.Info("Installing plugin")
 	err = pl.Register()
 	if err != nil {
+		log.Error(err.Error())
 		return err
 	}
 
 	reg, err = pl.IsRegistered(pInfo.Key)
 	if err != nil {
+		log.Error(err.Error())
```
Comment (jzt) [inline installer/fileserver/tasks/plugin.go]: `log.Error(err.Error())` is redundant, you can just use `log.Error(err)`.

File: installer/fileserver/tasks/plugin.go [hunk 7]
```diff
@@ -255,18 +200,21 @@ func (p *Plugin) Install(cli *cli.Context) error {
...
 		return err
 	}
 	if reg {
 		if p.Force {
 			log.Info("Removing existing plugin to force install")
 			err = pl.Unregister(pInfo.Key)
 			if err != nil {
+				log.Error(err.Error())
```
Comment (jzt) [inline installer/fileserver/tasks/plugin.go]: `log.Error(err)`

File: installer/fileserver/tasks/plugin.go [hunk 8]
```diff
@@ -255,18 +200,21 @@ func (p *Plugin) Install(cli *cli.Context) error {
 
 	pl, err := plugin.NewPluginator(context.TODO(), p.Target.URL, p.Target.Thumbprint, pInfo)
 	if err != nil {
+		log.Error(err.Error())
 		return err
 	}
 
 	reg, err := pl.IsRegistered(pInfo.Key)
 	if err != nil {
+		log.Error(err.Error())
```
Comment (jzt) [inline installer/fileserver/tasks/plugin.go]: `log.Error(err)`

File: installer/fileserver/tasks/plugin.go [hunk 9]
```diff
@@ -255,18 +200,21 @@ func (p *Plugin) Install(cli *cli.Context) error {
 
 	pl, err := plugin.NewPluginator(context.TODO(), p.Target.URL, p.Target.Thumbprint, pInfo)
 	if err != nil {
+		log.Error(err.Error())
```
Comment (jzt) [inline installer/fileserver/tasks/plugin.go]: here too

**Comment by jzt** [comment]
Lots of instances of `log.Error(err.Error())` that can be `log.Error(err)`. I'm not sure what those errors look like, but it may also make more sense to follow this pattern:  `log.Errorf("Error attempting to <do whatever the thing was>: %s", err)`

File: installer/fileserver/tasks/ova/configure.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+	host := ovaURL.Hostname()
+
+	op.Debugf("Looking up host %s", host)
+	ips, err := net.LookupIP(host)
+	if err != nil {
+		return nil, errors.Errorf("IP lookup failed: %s", err)
+	}
+
+	op.Debugf("found %d IP(s) from hostname lookup on %s:", len(ips), host)
```
Comment (andrewtchin) [inline installer/fileserver/tasks/ova/configure.go]: Maybe log the whole list?

File: installer/lib/login.go [hunk 11]
```diff
@@ -16,65 +16,91 @@ package lib
...
-	input.Password = &passwd
+	op.Infof("server URL: %v\n", info.URL.Host)
 
 	ctx, cancel := context.WithTimeout(context.Background(), loginTimeout)
 	loginResponse := make(chan error, 1)
-	var v *validate.Validator
 	var err error
 
 	go func() {
-		v, err = validate.NewValidator(ctx, input)
```
Comment (morris-jason) [inline installer/lib/login.go]: There's really no reason to use Validator here - we don't need any of the Network or Resource path verification. Instead, we can just use the underlying session verification.

File: installer/fileserver/tasks/plugin.go [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+		return err
+	}
+
+	// Do not attempt to install the Flex client plugin on vSphere 6.7
+	vCenterVersion := p.Target.Session.Client.ServiceContent.About.Version
+	if p.Key == flexClientPluginKey && vCenterVersion == vSphere67 {
+		err := errors.Errorf("Refusing to install Flex plugin on vSphere %s", vCenterVersion)
+		op.Error(err)
+		return err
+	}
```
Comment (morris-jason) [inline installer/fileserver/tasks/plugin.go]: @zjs does this address https://github.com/vmware/vic-ui/issues/452?

Comment (zjs) [reply installer/fileserver/tasks/plugin.go]: Can we make this a bit more future proof by looking for a version of 6.7 _or higher_? (That way, we wont "regress" if we forget to update this when the next version ships.)

Comment (morris-jason) [reply installer/fileserver/tasks/plugin.go]: @zjs added unit tests:
```
$> go test -timeout 30s github.com/vmware/vic-product/installer/fileserver/tasks -v
=== RUN   TestFlexInstallWithBadVersions
--- PASS: TestFlexInstallWithBadVersions (0.00s)
=== RUN   TestH5InstallWithAnyVersion
--- PASS: TestH5InstallWithAnyVersion (0.00s)
PASS
ok  	github.com/vmware/vic-product/installer/fileserver/tasks	0.018s
```

**Pull request merged**